doc/developer/design/20231102_read_holds.md

```diff
@@ -0,0 +1,156 @@
+It cannot exist on sinks or unmaterialized views.
+If a `HOLD` exists for a directly queryable object (i.e., not indexes) at time `t`, `SELECT` and `SUBSCRIBE` on that object `AS OF t` will successfully execute.
+If a `HOLD` exists for an index, the same is true if the index satisfies the query.
+A `HOLD` on any non-index object will not also create the `HOLD` on any indexes of the object.
+An `AS OF` query will restrict the indexes or object it considers to those that are valid for the requested time.
+
+Because this may be surprising to users, if an `AS OF` query whose plan uses an index errors because the `AS OF` is less than the index's read frontier and a `HOLD` exists on the index's object,
+a helpful error message will be produced telling the user that their `HOLD` must be on the index, not the underlying object.
+
+The objects the `HOLD` is created on do not need to be in the same schema or database.
+The `HOLD` can be created in any schema on which the user has a `CREATE` privilege (see (Alternatives)[#Alternatives] for other takes on this).
+The user must have a `SELECT` privilege on all objects.
+
+Terminology side note:
+Both the compute and storage controllers expose a `CollectionState` with `read_capabilities` and `implied_capability`.
+Although reading these are both exposed in the API, `read_capabilities` should never be read by the adapter, which should limit itself to reading only `implied_capability`.
+When this document uses the terms "read frontier" or "since", it is referring to the `implied_capability`.
+
+If `AT <time>` is specified, the read hold is created at that time.
+For each object in the `HOLD`, the time must be in advance of that object's read frontier *or* any other `HOLD`'s time on that object, otherwise an error occurs.
+We want to support safe swapping out of one `HOLD` for another, and since we only reference a collection's `implied_capability` (which is not aware of any read holds),
+it is necessary to call out specifically that other `HOLD`s times are valid for new `HOLD`s.
+If `AT <time>` is not specified, Materialize automatically chooses the least valid read frontier among all objects of the `HOLD`.
+
+`HOLD`s are destroyed by:
+
+```
+DROP HOLD <name>
+```
+
+A `DROP` will release its current read hold.
+In-progress `SUBSCRIBE`s and transactions will not be affected after a `DROP`, but will continue to run as expected.
+Those will still be aquiring their own read holds in addition to the `HOLD` object.
+
+Creating a `HOLD` will increase resource consumption (mostly memory) in proportion to the amount of time it is held for.
+Users must periodically advance the `HOLD` to a more recent time to allow for resources to be released.
+
+`HOLD`s are advanced by:
+
+```
+ALTER HOLD <name> ADVANCE [TO <time>]
+```
+
+If `TO <time>` is specified, the new time will be created at that time.
+It is an error if that time is less than any object's logical read frontier or less than the existing `HOLD`'s time.
```

Comment:
Does it need to be an error for a hold to go backwards in time? It seems plausibly fine, as long as there was something else holding open the previous time—maybe another hold, or the logical compaction policy on th eobject.

Reply:
Sure. Added words about this.